### PR TITLE
Lb/tracking by matching

### PR DESCRIPTION
**Pull request by LeonidBeynenson**
### This pullrequest changes

This pullrequest adds a tracker that uses Tracking-by-Matching approach.

The approach uses two different appearance measures to compute affinity between bounding boxes:
some fast descriptor and some strong descriptor. Each time the assignment
problem is solved. The assignment problem in our case is how to establish
correspondence between existing tracklets and recently detected objects.
First step is to compute an affinity matrix between tracklets and
detections. The affinity equals to
       appearance_affinity * motion_affinity * shape_affinity.
Where appearance is 1 - distance(tracklet_fast_dscr, detection_fast_dscr).
Second step is to solve the assignment problem using Kuhn-Munkres
algorithm. If correspondence between some tracklet and detection is
established with low confidence (affinity) then the strong descriptor is
used to determine if there is correspondence between tracklet and detection.

### CODE REVIEW

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+    /// \return Descriptor size.
+    ///
+    virtual cv::Size size() const = 0;
+
+    ///
+    /// \brief Computes image descriptor.
+    /// \param[in] mat Color image.
+    /// \param[out] descr Computed descriptor.
+    ///
+    virtual void Compute(const cv::Mat &mat, cv::Mat *descr) = 0;
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: This was Google style of code: output values are pointers to make the users write ampersands, to make them be sure to know what they do.

Is it possible to keep pointers here?

[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: It was Google style conventions.

Is it possible to keep Google style here?

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+    /// \param[in] descrs1 Batch of first descriptors.
+    /// \param[in] descrs2 Batch of second descriptors.
+    /// \return Distances between descriptors.
+    ///
+    std::vector<float> Compute(
+        const std::vector<cv::Mat> &descrs1,
+        const std::vector<cv::Mat> &descrs2) override;
+
+private:
+    cv::Size descriptor_size_;
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: I thought about that, but it seemed to me that these classes are too small for such factories

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alalek) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: As you wish.
OpenCV bindings probably would not work well here.

[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: I've replaced pointers by references for output parameters everywhere.

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+    /// \param[in] mats Images of interest.
+    /// \param[out] descrs Matrices to store the computed descriptors.
+    ///
+    virtual void Compute(const std::vector<cv::Mat> &mats,
+                         std::vector<cv::Mat> *descrs) = 0;
+
+    ///
+    /// \brief Prints performance counts for CNN-based descriptors
+    ///
+    virtual void PrintPerformanceCounts() const {}
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: I'll better remove it

File: modules/tracking/src/tracking_by_matching.cpp [hunk 4]
```diff
@@ -0,0 +1,1354 @@
...
+#define TBM_CHECK(cond) CV_Assert(cond)
+
+#define TBM_CHECK_BINARY(actual, expected, op)  CV_Assert(actual op expected)
+
+#define TBM_CHECK_EQ(actual, expected) TBM_CHECK_BINARY(actual, expected, ==)
+#define TBM_CHECK_NE(actual, expected) TBM_CHECK_BINARY(actual, expected, !=)
+#define TBM_CHECK_LT(actual, expected) TBM_CHECK_BINARY(actual, expected, <)
+#define TBM_CHECK_GT(actual, expected) TBM_CHECK_BINARY(actual, expected, >)
+#define TBM_CHECK_LE(actual, expected) TBM_CHECK_BINARY(actual, expected, <=)
+#define TBM_CHECK_GE(actual, expected) TBM_CHECK_BINARY(actual, expected, >=)
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/src/tracking_by_matching.cpp]: Wow!

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 5]
```diff
@@ -0,0 +1,537 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: Done

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+    ///
+    /// \brief Computes image descriptor.
+    /// \param[in] mat Color image.
+    /// \param[out] descr Computed descriptor.
+    ///
+    virtual void Compute(const cv::Mat &mat, cv::Mat *descr) = 0;
+
+    ///
+    /// \brief Computes image descriptors in batches.
+    /// \param[in] mats Images of interest.
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: It depends on the implementation

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+    /// \param[out] descrs Matrices to store the computed descriptors.
+    ///
+    virtual void Compute(const std::vector<cv::Mat> &mats,
+                         std::vector<cv::Mat> *descrs) = 0;
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: Done

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: Removed

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+    /// \brief Computes images descriptors.
+    /// \param[in] mats Frames containing images of interest.
+    /// \param[out] descrs Matrices to store the computed descriptors.
+    //
+    void Compute(const std::vector<cv::Mat> &mats,
+                 std::vector<cv::Mat> *descrs) override  {
+        CV_Assert(descrs != nullptr);
+        descrs->resize(mats.size());
+        for (size_t i = 0; i < mats.size(); i++)  {
+            Compute(mats[i], &(descrs[i]));
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: I really like "clean" abstract interface classes... So, I'd stay it here, if it is possible

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+    ///
+    /// Default constructor.
+    ///
+    TrackerParams();
+};
+
+///
+/// \brief The Track struct describes tracks.
+///
+struct CV_EXPORTS Track {
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: Done

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+    /// \brief Pipeline parameters getter.
+    /// \return Parameters of pipeline.
+    ///
+    virtual const TrackerParams &params() const = 0;
+
+    ///
+    /// \brief Pipeline parameters setter.
+    /// \param[in] params Parameters of pipeline.
+    ///
+    virtual void set_params(const TrackerParams &params) = 0;
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: Done

File: modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/include/opencv2/tracking/tracking_by_matching.hpp]: Fixed to CamelCase

File: modules/tracking/src/kuhn_munkres.hpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/src/kuhn_munkres.hpp]: Done

File: modules/tracking/src/tracking_by_matching.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/src/tracking_by_matching.cpp]: Done

File: modules/tracking/samples/tracking_by_matching.cpp [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+        char c = (char) waitKey( 2 );
+        if (c == 'q')
+            break;
+        if (c == 'p')
+            paused = !paused;
+    }
+
+    double s = frame_counter / (time_total / getTickFrequency());
+    printf("FPS: %f\n", s);
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply modules/tracking/samples/tracking_by_matching.cpp]: Sorry, but this code was copied from `tracker.cpp` sample in the same folder.

**Comment by llschloesser** [comment]
@LeonidBeynenson Does this have associated documentation, paper, tutorial, or sample?

**Pull request merged**